### PR TITLE
raftstore: only persist merge target if the merge is known to be succeeded

### PR DESCRIPTION
**Pull request by NingLin-P**
### What is changed and how it works?

Issue Number: Close #12232 

What's Changed:

When destroying, only persist merge target if the merge is known to be succeeded

```commit-message
raftstore: only persist merge target if the merge is known to be succeeded
```

### Check List 

Tests 

- Integration test

### Release note 

```release-note
fix tikv panic and peer unexpected destroy due to fake merge target
```

### CODE REVIEW

File: tests/failpoints/cases/test_merge.rs [hunk 1]
```diff
@@ -1666,3 +1666,84 @@ fn test_merge_pessimistic_locks_propose_fail() {
...
+    // and should not persist the `merge_state`
+    pd_client.must_remove_peer(right.get_id(), new_peer(5, 5));
+    must_get_none(&cluster.get_engine(5), b"k3");
+
+    // so that other peers will send message to store 5
+    pd_client.must_add_peer(right.get_id(), new_peer(5, 6));
+    // but it is still in tombstone state due to the message filter
+    let state = cluster.region_local_state(right.get_id(), 5);
+    assert_eq!(state.get_state(), PeerState::Tombstone);
+    assert!(!state.has_merge_state(), "{:?}", state);
```
Comment (BusyJay) [inline tests/failpoints/cases/test_merge.rs:1732]: Should probably be moved to L1745.

File: components/raftstore/src/store/peer.rs [hunk 2]
```diff
@@ -990,7 +990,12 @@ where
             &mut kv_wb,
             &region,
             PeerState::Tombstone,
-            self.pending_merge_state.clone(),
+            // Only persist the `merge_state` if the merge is known to be succeeded
+            // which is determined by the `keep_data` flag
+            self.pending_merge_state
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:995]: `if else` seems more readable.

**Review by BusyJay — APPROVED** [review]

**Review by 5kbpers — APPROVED** [review]

**Review by you06 — APPROVED** [review]

**Pull request merged**